pkg/gardenlet/controller/managedseed/managedseed_valueshelper.go

```diff
@@ -0,0 +1,362 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedseed
+
+import (
+	"io/ioutil"
+	"os"
+	"strings"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	"github.com/gardener/gardener/pkg/features"
+	"github.com/gardener/gardener/pkg/gardenlet/apis/config"
+	confighelper "github.com/gardener/gardener/pkg/gardenlet/apis/config/helper"
+	configv1alpha1 "github.com/gardener/gardener/pkg/gardenlet/apis/config/v1alpha1"
+	gardenletfeatures "github.com/gardener/gardener/pkg/gardenlet/features"
+	"github.com/gardener/gardener/pkg/utils"
+	"github.com/gardener/gardener/pkg/utils/imagevector"
+	"github.com/gardener/gardener/pkg/utils/secrets"
+
+	"github.com/Masterminds/semver"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/component-base/version"
+)
+
+// ValuesHelper provides methods for merging GardenletDeployment and GardenletConfiguration with parent,
+// as well as computing the values to be used when applying the gardenlet chart.
+type ValuesHelper interface {
+	// MergeGardenletDeployment merges the given GardenletDeployment with the values from the parent gardenlet.
+	MergeGardenletDeployment(*seedmanagementv1alpha1.GardenletDeployment, *gardencorev1beta1.Shoot) (*seedmanagementv1alpha1.GardenletDeployment, error)
+	// MergeGardenletConfiguration merges the given GardenletConfiguration with the parent GardenletConfiguration.
+	MergeGardenletConfiguration(config *configv1alpha1.GardenletConfiguration) (*configv1alpha1.GardenletConfiguration, error)
+	// GetGardenletChartValues computes the values to be used when applying the gardenlet chart.
+	GetGardenletChartValues(*seedmanagementv1alpha1.GardenletDeployment, *configv1alpha1.GardenletConfiguration, string, bool) (map[string]interface{}, error)
+}
+
+// valuesHelper is a concrete implementation of ValuesHelper
+type valuesHelper struct {
+	config      *config.GardenletConfiguration
+	imageVector imagevector.ImageVector
+}
+
+// newValuesHelper creates a new ValuesHelper with the given parent GardenletConfiguration and image vector.
+func newValuesHelper(config *config.GardenletConfiguration, imageVector imagevector.ImageVector) ValuesHelper {
+	return &valuesHelper{
+		config:      config,
+		imageVector: imageVector,
+	}
+}
+
+// MergeGardenletDeployment merges the given GardenletDeployment with the values from the parent gardenlet.
+func (vp *valuesHelper) MergeGardenletDeployment(deployment *seedmanagementv1alpha1.GardenletDeployment, shoot *gardencorev1beta1.Shoot) (*seedmanagementv1alpha1.GardenletDeployment, error) {
+	// Convert deployment object to values
+	deploymentValues, err := utils.ToValuesMap(deployment)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get parent deployment values
+	parentDeployment, err := getParentGardenletDeployment(vp.imageVector, shoot)
```

Comment:
It does merge **some** of the parent deployment parameters, those that the gardenlet code has access to at this point: the image, the `APIServerSNI` pod annotations, and the image vector overwrites (not in this method but still controlled by `MergeWithParent`). It wouldn't be a big deal for gardenlet to know some of the others as well (via the downward API), but I thought I should not do it unless there is an actual need for this. 

See above regarding my original intention with introducing `MergeWithParent`.

Reply:
After clarifying what `MergeWithParent` means, `getParentGardenletDeployment` should always be called and merged with, independently from `MergeWithParent`, see above. Perhaps also the code should be refactored to avoid implying that a merge with a "parent gardenlet deployment" is being done, since this is not really the case, as pointed out above.

Reply:
Done. Did not refactor the code, since after some experimenting I decided that producing a parent `GardenletDeployment` (with image and annotations) and merging with it is a very straightforward way to achieve what we want, and couldn't think of anything better. So I just made sure that this is done independently of whether `MergeWithParent` is true or not.